src/main/java/org/spongepowered/common/data/contextual/ContextualDataDelegate.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.contextual;
+
+import com.google.common.collect.Maps;
+import net.minecraft.network.protocol.Packet;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.api.data.DataHolder;
+import org.spongepowered.api.data.DataPerspective;
+import org.spongepowered.api.data.DataPerspectiveResolver;
+import org.spongepowered.api.data.value.Value;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.api.scoreboard.Team;
+import org.spongepowered.api.world.World;
+import org.spongepowered.plugin.PluginContainer;
+
+import java.util.Map;
+import java.util.Objects;
+
+public final class ContextualDataDelegate implements ContextualData {
+
+    private final DataPerspective perspective;
+    private final Map<DataPerspective, PerspectiveContainer<?, ?>> perspectives;
+
+    public ContextualDataDelegate(final DataPerspective perspective) {
+        this.perspective = perspective;
+
+        this.perspectives = Maps.newHashMap();
+    }
+
+    @Override
+    public @Nullable PerspectiveContainer<?, ?> getDataPerception(final DataPerspective perspective) {
+        return this.perspectives.get(perspective);
+    }
+
+    @Override
+    public PerspectiveContainer<?, ?> createDataPerception(final DataPerspective perspective) {
+        return this.perspectives.computeIfAbsent(perspective, p -> {
+            if (p instanceof final Entity entity) {
```

Comment:
Maybe add a bridge to all the perspectives implement that returns the container? Feels cleaner than this if chain that may keep growing over time.

Reply:
Could be done but it can also be super confusing because the `PerspectiveContainer` have different behavior depending on is the entity itself the holder as in `ContextualDataHolder` or some other type like `Team` which uses `ContextualDataDelegate` that propagates the updates differently.